### ISSUE TITLE
ArchUnit doesn't scan src/main/java when using Java Modules (Java 11)

### ISSUE THREAD

**Issue by adem-oz** [TNG/ArchUnit#497]
We have an issue similar to the #321 

ArchUnit doesn't scan the src/main/java when we have a Java Module (Java 11).

I will attach a demo.zip file which contains the demo.

Scenario to reproduice:
There is 3 projects in the demo.zip file. 
mycompany-parent: basic maven parent project which contains the versions of the libraries used in the demo.
mycompany-project1: maven project which is not a Java module (doesn't contain module-info.java)
mycompany-project2: maven project which is a Java module.

mvn install the projects: mycompany-parent, mycompany-project1 and mycompany-project2.

The mycompany-project2 contains the archunit test which is scanning src/main/java but it cannot find the file "SecondService.java".

[demo.zip](https://github.com/TNG/ArchUnit/files/5747910/demo.zip)

What is strange is if I change the package to scan "be.toto" inside "Project2ArchitectureTest.java". It finds the file inside that package but when I put it back to "be.mycompany" it doesn't find the files inside that package.

When I delete the module-info.java, it also find successfully the files inside package "be.mycompany".

Thanks for your help :)

**Comment by codecholeric** [comment]
Thanks for reporting this! :+1: I think this is a different issue though, if I understand it correctly. And I think for your issue the example project could also be reduced even more? At least I don't see why you need `project1` at all, since there is no cross dependency at the moment and your problem also is unrelated to classes in `project1`? (i.e. the test is within `project2` and so are the classes you're missing in your test?)

I think it might be related to the module path, i.e. you probably only need `project2` to reproduce it, as long as you use Java Modules. However I'm a little puzzled why the loader behaves differently for package `be.mycompany.project2` and `be.toto`, that seems to make no sense :thinking: 

I'll try to look into it more and then come up with a fix! (and understand why `be.mycompany.project2` and `be.toto` behave differently, I'm really curious about that :wink:)

**Comment by adem-oz** [comment]
With pleasure :)

Indeed, the project1 is not necessary to reproduce the issue. It was part of another test scenario that I didn't mention to not confuse you :)

Second test scenario: 
add project1 dependency in project2 pom.xml
run the test inside project2

I saw that the file inside project1 was detected when scanning occurs on the jars (scan on package be.mycompany). 

The project1 doesn't have a module-info. So it's not in the module path normally. Then I suspected that there is a problem when mixing project which has a module-info and project that doesn't have it.

**Comment by codecholeric** [comment]
I think I got a step further to at least understand why ArchUnit behaves the observed way at the moment (i.e. if you scan `be.mycompany` no violation is reported, but if you scan `be.toto` it finds the violation).
As far as I see the modulepath scanning doesn't work reliably for custom modules :thinking: The difference for the two packages is, that `target/test-classes` contains a package `be.mycompany`, but not `be.toto`.

I think what happens is the following:

* Java 9 modulepath scanning misses the custom module
* but it tries to find `ClassLoader.getResources(packageAsResource)` as a backup
* for `be.toto` this works, because it will find the location in `target/classes`
* for `be.mycompany` this doesn't work, because `--patch-module` will add `target/test-classes` and `ClassLoader.getResources("/be/mycompany")` will then ONLY return `target/test-classes`, since it contains that folder and `--patch-module` overrides the resource location and hides the original one (which I find odd, but I guess the JDK devs don't -> http://mail.openjdk.java.net/pipermail/core-libs-dev/2018-July/054228.html)

I guess the solution will be to fix step one and correctly scan custom modules, so the backup through `ClassLoader.getResources(..)` does not need to catch this. I'll add some test project to ArchUnit to prevent regression in the future after I've fixed this...

**Comment by codecholeric** [comment]
I think I've implemented a fix in #506. No warranties that it will work for all sort of combinations of JVM module parameters afterwards :wink:, but the example project you provided seems to have worked like expected with it.
If you want you can test this fix by adding the following repository to your POM:

```
<repositories>
  <repository>
    <id>sonatype-snapshots</id>
    <url>https://oss.sonatype.org/content/repositories/snapshots</url>
  </repository>
</repositories>
```

Then upgrade the ArchUnit version to `0.16.0-SNAPSHOT` to pull in a SNAPSHOT version that contains the fix.
I would be very interested to hear, if the fix feels good to you and solves all your problems! (or vice versa, if there is another issue left to fix)

**Comment by adem-oz** [comment]
Indeed, it works well in my projects as well :)

I guess it will be part of the next release. When is the next Release planned?

**Comment by codecholeric** [comment]
Kewl :tada: I will release this quickly then, basically merge #503 (where I'm currently waiting on a review), then release :+1:

**Comment by adem-oz** [comment]
thank you :-)
